File: backend/web/fetch_batch.go
package web

import (
	"context"
	"sync"
)

// FetchURLs extracts readable text from one or more http(s) URLs (parallel, max 3).
func fetchURLsParallel(ctx context.Context, urls []string) []PageHit {
	if len(urls) == 0 {
		return nil
	}
	if len(urls) > maxFetchCap {
		urls = urls[:maxFetchCap]
	}

	results := make([]PageHit, len(urls))
	var wg sync.WaitGroup
	sem := make(chan struct{}, 3)

	for i, rawURL := range urls {
		wg.Add(1)
		go func(i int, rawURL string) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()
			results[i] = fetchPage(ctx, rawURL)
		}(i, rawURL)
	}
	wg.Wait()
	return results
}

func fetchPage(ctx context.Context, rawURL string) PageHit {
	page, err := FetchPage(ctx, rawURL)
	if err != nil {
		var fe *FetchError
		if e, ok := err.(*FetchError); ok {
			fe = e
		} else {
			fe = &FetchError{Kind: FetchNetwork, Message: err.Error()}
		}
		return PageHit{URL: rawURL, Fetch: fe}
	}
	return page
}

File: backend/web/fetch_batch.ts
// PORT: backend/web/fetch_batch.go

import { Effect } from "effect";
import { type page_hit, fetch_error, fetch_failure_kind, max_fetch_cap } from "./types";
import { fetch_page } from "./fetch";

export const fetch_urls_parallel = (
  ctx: AbortSignal,
  urls: string[],
): Effect.Effect<page_hit[], never> => {
  if (urls.length === 0) {
    return Effect.succeed([]);
  }

  const limitUrls = urls.slice(0, max_fetch_cap);

  const fetchTasks = limitUrls.map((rawURL) => {
    return fetch_page(ctx, rawURL).pipe(
      Effect.catchAll((err) => {
        let fe: fetch_error;
        if (err instanceof fetch_error) {
          fe = err;
        } else {
          fe = new fetch_error(
            fetch_failure_kind.fetch_network,
            0,
            err instanceof Error ? err.message : String(err),
          );
        }
        const partial = fe.partialHit ?? { title: "", url: rawURL, content: "", fetch: null };
        return Effect.succeed({
          title: partial.title,
          url: partial.url,
          content: partial.content,
          fetch: fe,
        });
      }),
    );
  });

  return Effect.all(fetchTasks, { concurrency: 3 });
};

/*
PORT STATUS
source path: backend/web/fetch_batch.go
source lines: 47
confidence: high
status: phase_b_compile
*/
